File: QuadClaudeMac/QuadClaudeMac/Overlay/FileTreeNode.swift
import Foundation

/// Model for a node in the file explorer tree.
class FileTreeNode {
    static let excludedDirs: Set<String> = [
        ".git", "node_modules", "bin", "obj", ".vs", "__pycache__",
        ".next", "dist", "coverage", ".cache", ".nuget", "packages",
        ".terraform", "vendor", "target", "build", ".idea"
    ]

    static func isExcludedDir(_ name: String) -> Bool {
        return excludedDirs.contains(name) || name.hasPrefix(".")
    }

    let name: String
    let fullPath: String
    let isDirectory: Bool
    var children: [FileTreeNode] = []
    var isExpanded = false
    private var isLoaded = false

    var icon: String {
        if isDirectory {
            return isExpanded ? "\u{1F4C2}" : "\u{1F4C1}"
        }
        return Self.fileIcon(for: name)
    }

    init(fullPath: String, isDirectory: Bool) {
        self.fullPath = fullPath
        self.name = (fullPath as NSString).lastPathComponent
        self.isDirectory = isDirectory
    }

    func expand() {
        guard isDirectory, !isLoaded else { return }
        isExpanded = true
        loadChildren()
    }

    func collapse() {
        isExpanded = false
    }

    func loadChildren() {
        guard !isLoaded else { return }
        isLoaded = true
        children.removeAll()

        let fm = FileManager.default
        guard let contents = try? fm.contentsOfDirectory(atPath: fullPath) else { return }

        var dirs: [FileTreeNode] = []
        var files: [FileTreeNode] = []

        for item in contents {
            let itemPath = (fullPath as NSString).appendingPathComponent(item)
            var isDir: ObjCBool = false
            guard fm.fileExists(atPath: itemPath, isDirectory: &isDir) else { continue }

            if isDir.boolValue {
                if Self.isExcludedDir(item) { continue }
                dirs.append(FileTreeNode(fullPath: itemPath, isDirectory: true))
            } else {
                files.append(FileTreeNode(fullPath: itemPath, isDirectory: false))
            }
        }

        // Sort: folders first, then alphabetical
        dirs.sort { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
        files.sort { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
        children = dirs + files
    }

    func refresh() {
        guard isDirectory else { return }
        isLoaded = false
        children.removeAll()
        if isExpanded { loadChildren() }
    }

    // MARK: - File Icons

    private static func fileIcon(for name: String) -> String {
        let ext = (name as NSString).pathExtension.lowercased()
        switch ext {
        case "swift":           return "\u{1F426}"  // bird for Swift
        case "cs":              return "\u{2660}"    // spade
        case "ts", "tsx":       return "\u{25C6}"    // diamond
        case "js", "jsx":       return "\u{25CB}"    // circle
        case "json":            return "{ }"
        case "md":              return "\u{00B6}"    // pilcrow
        case "sh", "bash", "zsh": return ">_"
        case "xml", "html":     return "</>"
        case "css", "scss":     return "#"
        case "png", "jpg", "gif", "svg", "ico": return "\u{25A3}"
        case "xcodeproj", "xcworkspace": return "\u{2726}"
        default:                return "\u{2022}"    // bullet
        }
    }

    /// Search for files matching a query within this node's directory
    static func search(in rootPath: String, query: String, maxResults: Int = 50) -> [FileTreeNode] {
        let fm = FileManager.default
        var results: [FileTreeNode] = []

        guard let enumerator = fm.enumerator(
            at: URL(fileURLWithPath: rootPath),
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        ) else { return [] }

        while let url = enumerator.nextObject() as? URL {
            guard results.count < maxResults else { break }

            let name = url.lastPathComponent

            // Skip excluded directories
            if let isDir = try? url.resourceValues(forKeys: [.isDirectoryKey]).isDirectory, isDir == true {
                if isExcludedDir(name) {
                    enumerator.skipDescendants()
                    continue
                }
                continue
            }

            // Match file name
            if name.localizedCaseInsensitiveContains(query) {
                results.append(FileTreeNode(fullPath: url.path, isDirectory: false))
            }
        }

        return results
    }
}

File: QuadClaude/Overlay/FileTreeNode.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;

namespace QuadClaude.Overlay;

public class FileTreeNode : INotifyPropertyChanged
{
    private static readonly HashSet<string> ExcludedDirs = new(StringComparer.OrdinalIgnoreCase)
    {
        ".git", "node_modules", "bin", "obj", ".vs", "__pycache__",
        ".next", "dist", "coverage", ".cache", ".nuget", "packages",
        ".terraform", "vendor", "target", "build", ".idea"
    };

    public static bool IsExcludedDir(string name)
        => ExcludedDirs.Contains(name) || name.StartsWith('.');

    private bool _isExpanded;
    private bool _isLoaded;

    public string Name { get; }
    public string FullPath { get; }
    public bool IsDirectory { get; }
    public ObservableCollection<FileTreeNode> Children { get; } = [];

    public bool IsExpanded
    {
        get => _isExpanded;
        set
        {
            if (_isExpanded == value) return;
            _isExpanded = value;
            OnPropertyChanged();

            if (value && !_isLoaded && IsDirectory)
                LoadChildren();
        }
    }

    public string Icon => IsDirectory
        ? (_isExpanded ? "\uD83D\uDCC2" : "\uD83D\uDCC1")
        : GetFileIcon(Name);

    public FileTreeNode(string fullPath, bool isDirectory)
    {
        FullPath = fullPath;
        Name = Path.GetFileName(fullPath);
        IsDirectory = isDirectory;

        // Add a placeholder child so the expander arrow shows
        if (isDirectory)
            Children.Add(CreatePlaceholder());
    }

    private static FileTreeNode CreatePlaceholder()
    {
        return new FileTreeNode("", false)
        {
            _isLoaded = true
        };
    }

    public void LoadChildren()
    {
        if (_isLoaded) return;
        _isLoaded = true;

        Children.Clear();

        try
        {
            var entries = new List<FileTreeNode>();

            foreach (var dir in Directory.EnumerateDirectories(FullPath))
            {
                var dirName = Path.GetFileName(dir);
                if (ExcludedDirs.Contains(dirName)) continue;
                if (dirName.StartsWith('.')) continue; // skip hidden dirs
                entries.Add(new FileTreeNode(dir, true));
            }

            foreach (var file in Directory.EnumerateFiles(FullPath))
            {
                entries.Add(new FileTreeNode(file, false));
            }

            // Sort: folders first, then alphabetical
            entries.Sort((a, b) =>
            {
                if (a.IsDirectory != b.IsDirectory)
                    return a.IsDirectory ? -1 : 1;
                return string.Compare(a.Name, b.Name, StringComparison.OrdinalIgnoreCase);
            });

            foreach (var entry in entries)
                Children.Add(entry);
        }
        catch
        {
            // Permission denied or path too long — show empty
        }
    }

    public void Refresh()
    {
        if (!_isLoaded || !IsDirectory) return;
        _isLoaded = false;
        var wasExpanded = _isExpanded;
        Children.Clear();
        if (wasExpanded)
        {
            Children.Add(new FileTreeNode("Loading...", false) { _isLoaded = true });
            _isLoaded = false;
            LoadChildren();
        }
        else
        {
            Children.Add(new FileTreeNode("Loading...", false) { _isLoaded = true });
        }
    }

    private static string GetFileIcon(string name)
    {
        var ext = Path.GetExtension(name).ToLowerInvariant();
        return ext switch
        {
            ".cs" => "\u2660",      // spade — C#
            ".ts" or ".tsx" => "\u25C6", // diamond — TypeScript
            ".js" or ".jsx" => "\u25CB", // circle — JavaScript
            ".json" => "{ }",
            ".md" => "\u00B6",      // pilcrow — markdown
            ".sh" or ".bat" => ">_",
            ".xaml" or ".xml" or ".html" => "</>",
            ".css" or ".scss" => "#",
            ".png" or ".jpg" or ".gif" or ".svg" or ".ico" => "\u25A3", // square with fill
            ".sln" or ".csproj" => "\u2726", // four-pointed star
            _ => "\u2022"           // bullet
        };
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    private void OnPropertyChanged([CallerMemberName] string? name = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
        if (name == nameof(IsExpanded))
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Icon)));
    }
}
